`Pagination` - API refactoring

didoo opened the pull request:
### :pushpin: Summary

This PR collects all the work around the refactoring of the `Pagination` component API to make it more simple to adopt in the consumers codebase (see linked tickets).

For more details about why this change is needed, see the "More context" section below.

### :hammer_and_wrench: Detailed description

See commits for details.

Preview: https://hds-components-git-hds-972-pagination-cloud-ui-c491d7-hashicorp.vercel.app/components/pagination

### 🚧 Things to do:
- [x] how do we pass `@route/@model/@query` params to the numbered pages? 
   - after meeting and discussion with ambassadors, we've decided to try to explore this idea
   - this will be done in a follow-up PR (is not trivial)
- [x] ~~fix all the tests that are broken~~
   - ~~this will be done in a follow-up PR (to avoid too much noise)~~ 
   - https://github.com/hashicorp/design-system/pull/1050

### 🖇 Related tickets/PRs
- https://github.com/hashicorp/design-system/pull/661
- https://github.com/hashicorp/design-system/pull/1038
- https://github.com/hashicorp/design-system/pull/1050

Relevant thread
- https://hashicorp.slack.com/archives/C01KCU4HDPY/p1673471011040269

### 📒 More context

The problem that we're trying to solve with this PR arose once we tried to adopt the component (after #661) in Cloud UI, and we realized that our API (based on the idea of buttons triggering events) was not compatible with how the pagination was consumed in Cloud UI (based on `@route/@query` params provided to the controls).

So we had to find a way to "expose" the internals of the navigation controls, to the outside consuming codebase. We had different options, but we ended up deciding to use the contextual components within the `Nav` element.

<img width="1120" alt="screenshot_2201" src="https://user-images.githubusercontent.com/686239/211666319-668e95d5-45ed-43f2-907d-d608aba9f162.png">

This lead to this syntax:

```
  <Hds::Pagination as |P|>
    <P.Nav as |N|>
      <N.ButtonPrev @route="..." @query={{hash prev="previousPage" next=undefined}} />
      <N.ButtonNext @route="..." @query={{hash prev=undefined next="nextPage"}} />
    </P.Nav>
  </Hds::Pagination>
```

With this first attempt to allow passing dow the link parameters to the prev/next controls, we partially solved the problem, but not completely. How could we pass the routing parameters to the controls for the "page numbers" too? Plus, the resulting code (and components API) was quite convoluted at times (a lot of indirection and arguments passed around).
Plus, for the numbered pagination the parameters were passed to both the container as well as the contextual child components:

```
<Hds::Pagination @totalItems={{100}} @itemsPerPage={{30}} @currentPage={{3}} as |P|>
  <P.Info @showTotalItems={{false}} />
  <P.Nav @hideLabels={{false}} @isTruncated={{true}} />
  <P.SizeSelector @sizes={{array 10 30 50}} />
</Hds::Pagination>
```

The reason for this is that the previous implementation was trying to do too much in having a single pagination component handling two very different things (pagination with a known number of items, and with an unknown number of items) under the single component/code implementation.

So we decided to take another approach and opened a new PR (#1038) proposing to split the code in two different components:
- `Hds::Pagination::Numbered` - for use cases when the number of items is known, and this handles all the complexity of the pagination information shown to the user ("info", "page numbers", "current page", "page size") 
- `Hds::Pagination::Compact` - for use cases when the number of items is unknown, so it's impossible to have the "info", the "page numbers" and the "select-size" elements, and there is no concept of "current page"

This allowed us to bring the content of the `Pagination::Nav` directly in the `Compact` and `Numbered` component, removing one level of indirection, and make the `Pagination::Nav` not needed anymore. Same for the `Pagination` hybrid component, now replaced by two distinct/specialized components.

After discussing this approach in a dedicated meeting with the HDS ambassadors, we all came to the conclusion that the new proposed API (and overall approach) made sense, and was the right solution to the problem.

During the discussion the need to have "links" also for the "Numbered" pagination came up too (but this will be tackled in a follow-up PR).

***

### 👀 Reviewer's checklist:

- [ ] +1 Percy if applicable

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

KristinLBradley commented on packages/components/app/styles/components/pagination.scss:
```diff
@@ -49,6 +49,7 @@
 .hds-pagination-nav__control {
   // border width is subtracted from padding:
   padding: 0 calc(var(--token-pagination-nav-control-padding-horizontal) - 1px);
+  color: var(--token-color-foreground-primary);
```
I committed directly to this PR to add this line. I think I may have had the color as a CSS class previously and then deleted it accidentally in a later refactor? (Not sure.)

KristinLBradley commented:
For things to discuss #1:

To make the ButtonArrow components disabled in my test code, I used conditional logic so that an href value would not be provided. This makes the component render as an html button vs. an anchor tag so when the "disabled" attribute is applied it's valid.

```
<Hds::Pagination::Nav::ButtonArrow
    @direction="prev"
    @type={{this.type}}
    @disabled={{this.isDisabledPrev}}
    {{! Default - Uses button with onClick: }}
    @onClick={{unless @model (this.onPageChange)}}
    {{! Optional - Uses model with url query & route: }}
    @query={{if (and @model @model.previousPageToken) (hash next=undefined prev=@model.previousPageToken)}}
    @route={{if (and @model @model.previousPageToken) this.currentRouteName}}
/>
```

The syntax is a bit complex though.

KristinLBradley reviewed (commented):
(I'm going through individual commits, this feedback is one the first one)

The changes to "numbered", "compact", & "truncated" types seems simpler and more elegant and it allows some of the logic such as in the hbs to be simplified, so I'm happy with this update.

KristinLBradley commented on packages/components/addon/components/hds/pagination/nav/index.hbs:
```diff
@@ -1,7 +1,7 @@
   <Hds::Pagination::Nav::ButtonArrow
     @direction="prev"
-    @type={{this.type}}
+    @hideLabel={{eq this.type "numbered"}}
```
Thanks, this is a good update. The `type` I originally used here was actually tripping me up too since HTML buttons have their own `type` attribute.

KristinLBradley commented on packages/components/app/styles/components/pagination.scss:
```diff
@@ -70,10 +70,18 @@
     cursor: not-allowed;
+    // TODO discuss with Kristin if we want to do it (removes the cursor too!)
+    pointer-events: none;
   }
```
I added a comment to the PR with my test code as an example. I would prefer a solution which doesn't add an invalid attribute (the `disabled` attribute) to anchor link tags.

(Cloud-ui's Pagination also renders a button tag with the `disabled` attribute when the Prev & Next controls are disabled.)

alex-ju replied:
Not sure if it helps with the matter, but in my experience `<a role="link" aria-disabled="true">Link</a>` is one of the few solutions that convey a disabled link to screen reader users.

KristinLBradley replied:
Good to know. Would you need to add the role though since it's already a link? Or is that needed so that the screen reader both announces that it's a link AND that it's disabled.

It would still require extra work though as the link isn't truly disabled. (If a user clicks on it, a valid link would still work.)

alex-ju replied:
If the anchor doesn't have an href it is often treated as an inline element (similar to a span) which is why we're adding the role. The whole example is not needed anymore though, as we seem to be always rendering buttons for disabled states.

didoo replied:
This (the `pointer-events: none` and the disabled link) is outdated now, we are showing a disabled button. Resolving the conversation.

KristinLBradley reviewed (commented):
Looks good! Overall I agree with all your changes, the only thing I would change is how the disabled buttons are handled but that's relatively minor.

didoo commented:
> Looks good! Overall I agree with all your changes, the only thing I would change is how the disabled buttons are handled but that's relatively minor.

@KristinLBradley yes, you're right. I've changed the `ButtonArrow` control implementation. It's a bit move verbose, but it doesn't generate invalid code. See this commit:
https://github.com/hashicorp/design-system/pull/1020/commits/41d7c55c979a12757d589f3148145d1d9fa4986f

I've also fixed the tests too (I forgot 🤦‍♂️).

alex-ju commented on packages/components/addon/components/hds/pagination/nav/index.hbs:
```diff
@@ -1,10 +1,16 @@
-    @onClick={{this.onPageChange}}
-    @disabled={{this.isDisabledPrev}}
-  />
+  {{! TODO use `has-block-params`? https://api.emberjs.com/ember/4.4/classes/Ember.Templates.helpers/methods/action?anchor=has-block-params }}
+  {{#unless (has-block)}}
```
I think `has-block` is what we want here, as this would cover the cases where we have contextual components (`<N.ButtonPrev` and `N.ButtonNext`).

`<Hds::Pagination::Nav as |N|>` would return true for `has-block-params` but so will`<Hds::Pagination::Nav @totalPages={{10}} @currentPage={{3}} />`, so it's not probably what we want.

didoo replied:
Outdated.

alex-ju commented on packages/components/app/styles/components/pagination.scss:
```diff
@@ -70,23 +70,19 @@
-
+// only the "Previous" & "Next" (`ButtonArrow`) controls can be disabled
+.hds-pagination-nav__button-arrow {
+  // notice: when `@disabled={{true}}` is applied, the code always renders a `<button>` element
```
nice one! 👏

alex-ju reviewed (approved):
No suggestions for improvement from me. Enjoyed the granularity of the commits and the descriptions really helped with the review.

KristinLBradley commented on packages/components/app/styles/components/pagination.scss:
```diff
@@ -69,15 +70,19 @@
+// Make buttons with "Previous" & "Next" buttons the same min-width
 // to make them more visually centered in relationship to other sub-components
 .hds-pagination-nav__button-arrow--labeled {
   min-width: 103px;
```
Reminder: this is one we need to discuss again with @cveigt before final release.

didoo replied:
I am going to remove it for simplicity, we can always add it later if we see it's needed.

KristinLBradley commented on packages/components/tests/dummy/app/controllers/components/pagination.js:
```diff
@@ -9,4 +9,14 @@ export default class PaginationController extends Controller {
+
+  @action
+  handlePageChange(page) {
+    console.log(`Page changed to "${page}"!`);
```
Do you want to leave these console logs in for now?

didoo replied:
no, it's on purpose: it's to have something to show to the consumers, that they have clicked a button and triggered an event/callback, when they play with one of the examples in the "how to use" or "showcase"

didoo commented:
@KristinLBradley @alex-ju I have asked you to re-review this PR, because I've made **a lot** of changes and refactorings to the component. I can'd describe in detail all the changes, please have a look at the commit messages (I've tried to be clear in what I did) and at the overall changes.

Please review (and test) thoroughly, because I expect to have missed things. As mentioned above, the integration tests will be fixed in a follow-up PR (and then merged here once both the PRs are approved)

----

@MelSumner do you mind to do another pass on the component, to make sure it's at the level of accessibility we want? thanks

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -23,94 +18,316 @@
-  </p>
+    The term «<code class="dummy-code">Pagination</code>
+    component» is an umbrella term used to indicate a set of Pagination-related components and sub-components, each with
+    their own APIs.</p>
```
I would use regular English quotes: 
“”
vs.
«»
which I think of as being "French quotes" (may be used in other languages too but I took French in high school & college).

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -23,94 +18,316 @@
-    <code class="dummy-code">Pagination</code>
-    component is a wrapper component for the Pagination-related sub-components, each with their own APIs:
-  </p>
+    The term «<code class="dummy-code">Pagination</code>
```
```suggestion
    The term “<code class="dummy-code">Pagination</code>
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -23,94 +18,316 @@
-    component is a wrapper component for the Pagination-related sub-components, each with their own APIs:
-  </p>
+    The term «<code class="dummy-code">Pagination</code>
+    component» is an umbrella term used to indicate a set of Pagination-related components and sub-components, each with
```
```suggestion
    component” is an umbrella term used to indicate a set of Pagination-related components and sub-components, each with
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -23,94 +18,316 @@
-    <dt>totalItems <code>number</code> <strong class="required">required</strong></dt>
-    <dd>Pass the total number of items to be paginated. If no value is defined an error will be thrown.</dd>
+  <p class="dummy-paragraph">
+    These pagination sub-elements may be used directly if you need to cover a very specific use need that is not covered
```
```suggestion
    These pagination sub-elements may be used directly if you need to cover a very specific use case that is not covered
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -23,94 +18,316 @@
+        Example:
+        {{! prettier-ignore-start }}
+        {{! template-lint-disable no-unbalanced-curlies }}
+        <code>@pageSizes=\{{array 10 30 50}}</code>
```
Btw, I wondered if it would be useful to provide these values as default page sizes which could be overridden if desired.

didoo replied:
Yeah, I thought the same. Good point. Can you investigate this with @cveigt + the ambassadors, and see if it's possible to come up with two reasonable defaults for:
- `@pagesizes` (available options in the `<select>/Hds::Pagination::SizeSelect`)
- `@selectedSize` (selected option)
and in that case open a PR against this one?

didoo replied:
[Note: this is not a blocker for the release, because this change would not be a breaking change, so we may introduce it later as a _minor_ update]

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -23,94 +18,316 @@
+    <dd>
+      <p>Callback function invoked (if provided) when one of the navigation controls is clicked, and a "page" change is
+        triggered.</p>
+      <p>The function receives the value of the "go to" page as argument (integer number).</p>
```
```suggestion
      <p>The function receives the value of the "go to" page as an argument (integer number).</p>
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -23,94 +18,316 @@
+    <dd>
+      <p>The
+        <code>@route</code>
+        parameter passed down as arguments to internal render a
```
```suggestion
        parameter passed down as arguments to internally render a
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -23,94 +18,316 @@
+        spreading is supported on this component.</p>
+      <p><em>Notice: it will not be applied to the
+          <code class="dummy-code">&lt;li&gt;</code>
+          wrapping element but to the
```
```suggestion
          wrapping element but to the nested
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -124,20 +341,20 @@
+        to the
         <code class="dummy-code">Pagination::SizeSelector</code>
-        itself when used as a stand alone component</p>
+        component itself when used as a stand alone component</p>
```
```suggestion
        component itself when used as a stand alone component.</p>
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+    >cursor-based</a>
+    (<strong>"compact"</strong>).</p>
+  <p class="dummy-paragraph">In the first one, the user is presented with a list of navigation controls ("prev/next" and
+    "page numbers" to go directly to a specific page) and other optional UI elements; in the second, much simpler, the
```
```suggestion
    "page numbers" to go directly to a specific page) and other optional UI elements; in the second, much simpler one, the
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+    "page numbers" to go directly to a specific page) and other optional UI elements; in the second, much simpler, the
+    user is presented with only the "prev/next" controls.</p>
+
+  <p class="dummy-paragraph">When the pagination is invoked using directly one of these two components, it will
```
```suggestion
  <p class="dummy-paragraph">When pagination is invoked directly using one of these two components, it will
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+    </li>
+    <li>
+      <p class="dummy-paragraph">manage the "current page" status across the different sub-components it's made of
+        (based on the arguments provided to it and its children)</p>
```
```suggestion
        (based on the arguments provided to it and its children).</p>
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+      <p class="dummy-paragraph">the organization/layout of the elemens in the page</p>
+    </li>
+    <li>
+      <p class="dummy-paragraph">the logic to handle the "current page" status</p>
```
```suggestion
      <p class="dummy-paragraph">the logic to handle the "current page" status.</p>
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+    event that calls the
+    <code class="dummy-code">onPageChange</code>
+    function.</p>
+  <p class="dummy-paragraph">This means that if you need to update the URL when the user changes "page" in the
```
```suggestion
  <p class="dummy-paragraph">This means that if you need to update the URL when the user changes the "page" in the
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+    <code class="dummy-code">onPageChange</code>
+    function.</p>
+  <p class="dummy-paragraph">This means that if you need to update the URL when the user changes "page" in the
+    Pagination (eg. to add/remove/update some query parameters) you have to do it inside the
```
```suggestion
    Pagination (eg. to add/remove/update some query parameters), you have to do it inside the
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+    Pagination (eg. to add/remove/update some query parameters) you have to do it inside the
+    <code class="dummy-code">onPageChange</code>
+    callback you provide to the component.</p>
+  <p class="dummy-paragraph">If instead you need to update direcly the URL when the user clicks on one of the
```
```suggestion
  <p class="dummy-paragraph">If instead you need to update the URL directly when the user clicks on one of the
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+
+  <p class="dummy-paragraph">The basic invocation of the "numbered" pagination requires the
+    <code class="dummy-code">@totalItems</code>
+    <code class="dummy-code">@itemsPerPage</code>
```
```suggestion
    <code class="dummy-code">@itemsPerPage</code>,
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+  <h5 class="dummy-h5">Extra arguments</h5>
+  <p class="dummy-paragraph">It's possible to customize the "Info", "Controls", and "SizeSelector" providing additional
+    arguments to them. For more details about these parameters, refer to the "Components API" section.</p>
+  <p class="dummy-paragraph">Below an example of some of these extra arguments:</p>
```
```suggestion
  <p class="dummy-paragraph">Below is an example of some of these extra arguments:</p>
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
-
-  <h4 class="dummy-h4">“Numbered” pagination with all optional child components</h4>
+  <Hds::Pagination::Numbered @totalItems={{150}} @itemsPerPage={{10}} @pageSizes={{array 10 30 50}} />
+  <p class="dummy-paragraph">If it's necessary it's possible to disable this functionality using a specific
```
```suggestion
  <p class="dummy-paragraph">If necessary, it's possible to disable this functionality using a specific
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+    @onPageSizeChange={{this.handlePageSizeChange}}
+  />
+  <p class="dummy-paragraph">The first function
+    <code class="dummy-code">onPageChange</code>
```
I would change this to:

"The first function, `onPageChange`, is ..."

Or perhaps to:

"The first `onPageChange` function is ..."

didoo replied:
Fixed

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+    is invoked when a user interacts with a navigation control ("prev/next" or "page number") and so can be used to
+    respond to a "page" change (eg. updating the list of items in the page and/or updating the routing/URL).</p>
+  <p class="dummy-paragraph">The second function
+    <code class="dummy-code">onPageSizeChange</code>
```
Similar to above, I would change this to:

"The second function, `onPageSizeChange`, is ..."

Or perhaps to:

"The second `onPageSizeChange` function is ..."

didoo replied:
Fixed

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+  <p class="dummy-paragraph">The second function
+    <code class="dummy-code">onPageSizeChange</code>
+    is invoked when a user interacts with the "size selector" and so can be used to respond to a "page size" change (eg.
+    updating the number of items listed in the page, updating the routing/URL, updating other elements in the page).</p>
```
```suggestion
    updating the number of items listed in the page, updating the routing/URL, and/or updating other elements in the page).</p>
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+    updating the number of items listed in the page, updating the routing/URL, updating other elements in the page).</p>
+
+  <h5 class="dummy-h5">Routing/URL updates</h5>
+  <p class="dummy-paragraph">If you want the pagination to change directly the URL of the page (eg. adding/removing
```
```suggestion
  <p class="dummy-paragraph">If you want the pagination to change the URL of the page directly (eg. adding/removing
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+  {{! =============================== }}
+
+  <h4 class="dummy-h4">Pagination::Compact</h4>
+  <p class="dummy-paragraph">By default, the basic use of the pagination provides .</p>
```
```suggestion
  <p class="dummy-paragraph">By default, the basic use of the pagination provides:</p>
```

KristinLBradley replied:
Not sure if something was left out of the above.

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+  <h4 class="dummy-h4">Pagination::Compact</h4>
+  <p class="dummy-paragraph">By default, the basic use of the pagination provides .</p>
+
+  <p class="dummy-paragraph">The basic invocation of the "compact" pagination doesn't require any argument (apart from
```
```suggestion
  <p class="dummy-paragraph">The basic invocation of the "compact" pagination doesn't require any arguments (apart from
```

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
 
-  <h4 class="dummy-h4">“Truncated” PaginationNav</h4>
+  <h5 class="dummy-h5">Extra arguments</h5>
+  <p class="dummy-paragraph">If necessry, it's possible to hide the labels in the controls:</p>
```
```suggestion
  <p class="dummy-paragraph">If necessary, it's possible to hide the labels in the controls:</p>
```

didoo replied:
Fixed

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
-  }}
-</section>
+  <h5 class="dummy-h5">Events handling</h5>
+  <p class="dummy-paragraph">The component exposes a callback function that can be used to respond to :</p>
```
Missing some language? Perhaps change to below:

"respond to :"
=>
"respond to page changes:"

didoo replied:
Fixed

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+  <p class="dummy-paragraph">Renders to:</p>
+  <Hds::Pagination::Compact @onPageChange={{this.handlePageChange}} />
+  <p class="dummy-paragraph">The function
+    <code class="dummy-code">onPageChange</code>
```
I would change the order to:

The `onPageChange` function is ...

didoo replied:
Fixed

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+    "page" change (eg. updating the list of items in the page and/or updating the routing/URL).</p>
+
+  <h5 class="dummy-h5">Routing/URL updates</h5>
+  <p class="dummy-paragraph">If you want the pagination to change directly the URL of the page (eg. adding/removing
```
```suggestion
  <p class="dummy-paragraph">If you want the pagination to change the URL of the page directly (eg. adding/removing
```

didoo replied:
Outdated.

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -148,157 +365,295 @@
+
+  <h5 class="dummy-h5">Routing/URL updates</h5>
+  <p class="dummy-paragraph">If you want the pagination to change directly the URL of the page (eg. adding/removing
+    query parameters) you need to pass the routing parameters to the component:</p>
```
```suggestion
    query parameters), you need to pass the routing parameters to the component:</p>
```

didoo replied:
Outdated.

KristinLBradley reviewed (approved):
Looks great overall. I just added minor grammar and spelling fix suggestions.

didoo commented on packages/components/tests/dummy/app/templates/components/pagination.hbs:
```diff
@@ -23,94 +18,316 @@
+    </dd>
+    <dt>disabled <code>boolean</code></dt>
+    <dd>
+      <p>If the control is disabled</p>
```
Fixed

didoo commented:
> Looks great overall. I just added minor grammar and spelling fix suggestions.

Thanks for the thorough spell/grammar checking! 🙏

I think I have _fixed_ all the suggestions you made, but if I missed something let me know.

MelSumner commented on packages/components/addon/components/hds/pagination/nav/ellipsis.hbs:
```diff
@@ -1,3 +1 @@
-  <div class="hds-pagination-nav__ellipsis">...</div>
-</li>
\ No newline at end of file
+<div class="hds-pagination-nav__ellipsis" ...attributes>...</div>
```
Can you remind me how this would be handled for assistive tech?

didoo replied:
@MelSumner looking at other design systems, this is exactly how it's implemented in all of them:
- https://design-system.service.gov.uk/components/pagination/
- https://design-system.w3.org/components/pagination.html
- https://atlassian.design/components/pagination/examples
- https://ons-design-system.netlify.app/components/pagination/
- https://primer.style/css/components/pagination

What would you suggest to do?

alex-ju reviewed (approved):
Looks great! Sorry for the slow(er) review on this.

thiskevinwang commented:
Just offering up some observations (since I'm doing related work on the React side of things).
It appears that none of the `Pagination::Nav::Number` buttons are tab-focusable on Safari.

didoo commented:
> ~Just offering up some observations (since I'm doing related work on the React side of things). It appears that none of the `Pagination::Nav::Number` buttons are tab-focusable on Safari (`Version 16.0 (17614.1.25.9.10, 17614)`).~ 
> Edit 1: Disregard! I had to **Enable Full Keyboard Access** 

Exactly! I was going to share this link 😀
https://jansensan.net/technical/how-fix-focus-rings-not-showing-links-macos

didoo commented:
since the PR #1060 for the API refactoring (to support routing-based pagination) has been much more complex than was expected, with a lot of commits, doesn't make sense to merge it into this PR; so I've decided to merge this one on the `feature branch` and then open a separate PR (squashing and cleaning up the commits) for the API refactoring for routing
